lib/src/rewrite.rs

```diff
@@ -464,70 +464,107 @@ pub struct MoveCommitsStats {
     pub num_abandoned: u32,
 }
 
+pub enum MoveCommitsTarget {
+    /// The commits to be moved. Commits should be mutable and in reverse
+    /// topological order.
+    Commits(Vec<Commit>),
+    /// The root commits to be moved, along with all their descendants.
+    /// Root commits should be mutable and in reverse topological order.
+    Roots(Vec<Commit>),
+}
+
 /// Moves `target_commits` from their current location to a new location in the
 /// graph.
 ///
-/// Commits in `target_roots` are rebased onto the new parents
-/// given by `new_parent_ids`, while the `new_children` commits are
-/// rebased onto the heads of `target_commits`. If `target_roots` is
-/// `None`, it will be computed as the roots of the connected set of
-/// target commits. This assumes that `target_commits` and
-/// `new_children` can be rewritten, and there will be no cycles in
-/// the resulting graph. `target_commits` should be in reverse
-/// topological order. `target_roots`, if provided, should be a subset
-/// of `target_commits`.
+/// Commits in `target` are rebased onto the new parents given by
+/// `new_parent_ids`, while the `new_children` commits are rebased onto the
+/// heads of the commits in `targets`. This assumes that commits in `target` and
+/// `new_children` can be rewritten, and there will be no cycles in the
+/// resulting graph. Commits in `target` should be in reverse topological order.
 pub fn move_commits(
     settings: &UserSettings,
     mut_repo: &mut MutableRepo,
     new_parent_ids: &[CommitId],
     new_children: &[Commit],
-    target_commits: &[Commit],
-    target_roots: Option<&[CommitId]>,
+    target: MoveCommitsTarget,
     options: &RebaseOptions,
 ) -> BackendResult<MoveCommitsStats> {
-    if target_commits.is_empty() {
-        return Ok(MoveCommitsStats {
-            num_rebased_targets: 0,
-            num_rebased_descendants: 0,
-            num_skipped_rebases: 0,
-            num_abandoned: 0,
-        });
-    }
-
-    let target_commit_ids: HashSet<_> = target_commits.iter().ids().cloned().collect();
+    let target_commits: Vec<Commit>;
+    let target_commit_ids: HashSet<_>;
+    let connected_target_commits: Vec<Commit>;
+    let connected_target_commits_internal_parents: HashMap<CommitId, Vec<CommitId>>;
+    let target_roots: HashSet<CommitId>;
+
+    match target {
+        MoveCommitsTarget::Commits(commits) => {
+            if commits.is_empty() {
+                return Ok(MoveCommitsStats {
+                    num_rebased_targets: 0,
+                    num_rebased_descendants: 0,
+                    num_skipped_rebases: 0,
+                    num_abandoned: 0,
+                });
```

Comment:
nit: `#[derive(Default)]` could be added to `MoveCommitStats` to make this shorter.